stats/stats.py

```diff
@@ -0,0 +1,403 @@
+
+                    # if the selected modifier is incorrect
+                    elif self.modifier not in values['modifiers']:
+                        raise ValueError('Invalid modifier')
+
+                # foreach granulation field verify if it has granulation parameters or if they
+                # are correct
+                for granulation_field, granulation_parameters in values['granulations'].iteritems():
+                    self._validate_granulation_parameters(granulation_field, granulation_parameters,
+                                                          values['granulations'])
+
+        return True
+
+    # iterates the granulated_stats dictionary and creates the stats_list
+    def _create_stats_list(self, granulated_stats, stats_list, granulation_field,
+                           granulation_arguments):
+        for key in sorted(granulated_stats.keys()):
+            data = dict()
+            data['currency'] = key.currency
+            if 'additional_granulation_field' in granulation_arguments:
+                data['customer_name'] = key.name
+            if 'granulation_field' in granulation_arguments:
+                data[granulation_field] = key.granulation
+            data['values'] = []
+            for subscription in sorted(granulated_stats[key]):
+                data['values'].append(subscription)
+            stats_list.append(data)
+
+    # search the elements that correspond with the granulation tuple key given
+    def _search_element(self, tuple_key, granulated_stats, granulation_arguments, customer_name,
+                        amount, id, additional_field):
+
+        search_tuple_key = granulated_stats.get(tuple_key, [])
+
+        details = {'id': id}
+        if amount is not None:
+            details['total'] = amount
+        if 'additional_granulation_field' not in granulation_arguments:
+            details['customer_name'] = customer_name
+        if additional_field is not None:
+            details[additional_field.items()[0][0]] = additional_field.items()[0][1]
+
+        granulated_stats[tuple_key] = search_tuple_key + [details]
+
+    # names the keys in the granulation tuple thus providing a way to access them
+    def _name_tuple(self, granulation_arguments):
+        if 'granulation_field' in granulation_arguments:
+            if 'additional_granulation_field' in granulation_arguments:
+                return namedtuple("group_by", ["name", "granulation", "currency"])
+            else:
+                return namedtuple("group_by", ["granulation", "currency"])
+        elif 'additional_granulation_field' in granulation_arguments:
+            return namedtuple("group_by", ["name", "currency"])
+        else:
+            return namedtuple("group_by", ["currency"])
+
+    # gets a tuple formed by the selected granulations
+    def get_granulation_tuple(self, granulation_arguments, customer_name, currency, amount,
```

Comment:
Try not to pass each field that's contained in the stats separately, such as `amount`, `currency`, `customer`... Maybe put them in a dict and pass them like that.